### DISCUSSION TITLE
How to count on the last item in one-to-many relation?

### DISCUSSION THREAD

**Discussion by turakvlad** [prisma/prisma#10826] [Q&A] [answered]
Hello! 

I found a similar question to mine here – https://github.com/prisma/prisma/discussions/4298. However, I think, my issue is a bit simpler and maybe it would be possible to solve it with Prisma instead of a raw query. 

Assume that we have 2 models:

1. `Certificate` model
2. `EmailTracking` model (it has `isDelivered`, `isOpened`, `isDownloaded`, etc)

One `Certificate` may have many `EmailTrackings`. A User can publish a `Certificate` (this action creates `EmailTracking`) and later send a few more follow-ups to the same recipient. Each "send" action creates an `EmailTracking`. So, one `Certificate` may have multiple `EmailTrackings`.

What I need is to count unique opened `Certificates` based on **last** `EmailTracking`. 

If I wrote a query like this:

```
      prisma.certificate.count({
        where: {
          emailTrackings: {
            some: {
              isOpened: true
            }
          }
        }
      }),
```

Prisma will return a number of `Certificates` that have at least one `EmailTracking` with `isOpened=true`. I cannot use `some` since there is a chance when the first email was opened, a User resent a `Certificate` again (new `EmailTracking` has been created), and this second email remains unopened. Our analytics should show certificates based on last `EmailTracking`.

So, I need something like:

```
      prisma.certificate.count({
        where: {
          emailTrackings: {
            last: { // note this "last" here (sorted by ID).
              isOpened: true
            }
          }
        }
      }),
```

Prisma [relation filters](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#relation-filters) do not have something like `last`.

Is there any way to achieve what I want with Prisma? Maybe I can use `orderBy` somehow to order with `ID=desc` or something like that?

Thanks in advance!

**Comment by austincrim** [comment, marked as answer]
Hey @turakvlad 👋 ,

I don't think this is possible with a Prisma query, though it most likely can be done with SQL and `$queryRaw`.

If you would like to see us add something like this, feel free to create a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with your use case.

Thanks!

**Reply by turakvlad** [reply]
Thank you! I have just opened a feature request here – https://github.com/prisma/prisma/issues/11109.

**Reply by turakvlad** [reply]
Hey @austincrim :) 

I have one more question. Let's say I will write this query by myself. Is there any way I can extend Prisma [relation filters](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#relation-filters) with this `last` functionality?

In my `where` input I use a lot of build-in Prisma features. My `where` object looks like:

```
{
  OR: [
    {},
    {}
  ],
  group: {
    id: {
      in: []
    }
  },
  isPublished: true | false,
  emailTrackings: {
    some: { isOpened: true } // and here I need to use my "last"
  }
}
```

If I need to write a raw query from scratch, I will need to move everything to that query, won't I? 

That's why I am asking whether it is possible either extend Prisma with `last` functionality or somehow to add my raw query for fetching last `emailTrackings` to existing `where` object? Otherwise, I will need to move everything a raw query.

I hope you understand my question :) Thank you!

**Reply by austincrim** [reply]
Hello again @turakvlad!

You will need to move everything to a raw query at this point.

Extending parts of the client is a pretty common request, so feel free to add your +1 or specific use case to the issue (#7161) to help our engineers prioritize and design the feature.

Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
